DigitalLearningSolutions.Web.Tests/Controllers/Register/RegisterInternalAdminControllerTests.cs

```diff
@@ -0,0 +1,286 @@
+        private HttpRequest request = null!;
+
+        [SetUp]
+        public void Setup()
+        {
+            centresDataService = A.Fake<ICentresDataService>();
+            centresService = A.Fake<ICentresService>();
+            userDataService = A.Fake<IUserDataService>();
+            registrationService = A.Fake<IRegistrationService>();
+            delegateApprovalsService = A.Fake<IDelegateApprovalsService>();
+            featureManager = A.Fake<IFeatureManager>();
+            registerAdminHelper = new RegisterAdminHelper(userDataService, centresDataService);
+            request = A.Fake<HttpRequest>();
+            controller = new RegisterInternalAdminController(
+                    centresDataService,
+                    centresService,
+                    userDataService,
+                    registrationService,
+                    delegateApprovalsService,
+                    featureManager,
+                    registerAdminHelper
+                )
+                .WithDefaultContext()
+                .WithMockRequestContext(request)
+                .WithMockUser(true, userId: DefaultUserId)
+                .WithMockServices()
+                .WithMockTempData();
+        }
+
+        [Test]
+        public void IndexGet_with_no_centreId_param_shows_notfound_error()
+        {
+            // When
+            var result = controller.Index();
+
+            // Then
+            result.Should().BeNotFoundResult();
+        }
+
+        [Test]
+        public void IndexGet_with_invalid_centreId_param_shows_notfound_error()
+        {
+            // Given
+            A.CallTo(() => centresDataService.GetCentreName(DefaultCentreId)).Returns(null);
+
+            // When
+            var result = controller.Index(DefaultCentreId);
+
+            // Then
+            A.CallTo(() => centresDataService.GetCentreName(DefaultCentreId)).MustHaveHappenedOnceExactly();
+            result.Should().BeNotFoundResult();
+        }
+
+        [Test]
+        public void IndexGet_with_allowed_admin_registration_returns_view_model()
+        {
+            // Given
+            A.CallTo(() => centresDataService.GetCentreName(DefaultCentreId)).Returns("Some centre");
+            A.CallTo(() => centresDataService.GetCentreAutoRegisterValues(DefaultCentreId))
+                .Returns((false, "email@email"));
+            A.CallTo(() => userDataService.GetAdminsByCentreId(DefaultCentreId)).Returns(new List<AdminEntity>());
+
+            // When
+            var result = controller.Index(DefaultCentreId);
+
+            // Then
+            A.CallTo(() => centresDataService.GetCentreName(DefaultCentreId)).MustHaveHappenedOnceExactly();
+            A.CallTo(() => centresDataService.GetCentreAutoRegisterValues(DefaultCentreId))
+                .MustHaveHappenedOnceExactly();
+            A.CallTo(() => userDataService.GetAdminsByCentreId(DefaultCentreId)).MustHaveHappenedOnceExactly();
```

Comment:
You've tested these methods in `RegisterAdminHelperTests`, so you can just mock `registerAdminHelper.IsRegisterAdminAllowed` to return true here

Reply:
There is no test that when `registerAdminHelper.IsRegisterAdminAllowed` is false, the AccessDenied page is returned

Reply:
Apparently only virtual members can be mocked, so I had to create an interface `IRegisterAdminHelper`, but I don't see any harm with that.

Reply:
I realise this convention isn't documented yet, but the difference between a helper class and a service class is that
* Helpers are static and contain only data and pure functions. These won't have interfaces, because they're static.
* Services are not static. They can make calls to the db. These will always have interfaces so they can be mocked.

From looking at RegisterAdminHelper, it's actually a service. Can you rename it?
Feel free to argue back if you disagree.

Reply:
Written a doc: https://softwiretech.atlassian.net/wiki/spaces/HEE/pages/20162936850/Static+helper+classes+v.s.+non-static+services+with+interfaces